packages/smart-contracts/src/contracts/ERC20EscrowToPayV1.sol

```diff
@@ -0,0 +1,385 @@
+/// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.7;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/TokenTimelock.sol";
+import "./interfaces/ERC20FeeProxy.sol";
+
+/**
+ * @title   ERC20EscrowToPayV1.
+ * @author  Ivo Garofalo.
+ * @notice  Request Invoice with Escrow.
+ */
+contract ERC20EscrowToPayV1 {
+    IERC20FeeProxy paymentProxy;
+    address payable owner;
+
+    struct Invoice {
+        IERC20 paymentToken; 
+        uint256 amount;
+        address payee;
+        address payer;
+        uint feeAmount;
+        address feeAddress;
+        uint256 claimDate;
+    }
+
+    struct Dispute {
+        bytes paymentReference;
+        IERC20 paymentToken; 
+        uint256 amount;
+        address payee;
+        address payer;
+        uint feeAmount;
+        address feeAddress;
+        uint256 duration;
+        uint256 endTime;
+        TokenTimelock tokentimelock;
+    }
+
+    mapping(bytes => Invoice) public invoiceMapping;
+    mapping(bytes => Dispute) public disputeMapping;
+
+    modifier OnlyOwner {
+        require(
+            msg.sender == owner,
+            "ERC20EscrowToPay: Only owner can call this function."
+        );
+        _;
+    }
+    modifier OnlyPayer(bytes memory _paymentRef) {
+        require(
+            msg.sender == disputeMapping[_paymentRef].payer,
+            "ERC20EscrowToPay: Only payer can call this function."
+        );
+        _;
+    }
+
+    /// Errors
+    error WithdrawFailed(string description);
+
+    /**
+     * @notice Emitted when an new escrow is initiated.
+     * @param paymentReference Reference of the payment related.
+     */
+    event OpenEscrow(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when an non-disputed escrow is completed successfully.
+     * @param paymentReference Reference of the payment related.
+     */
+    event EscrowCompleted(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when a new dispute is initiated.
+     * @param paymentReference Reference of the payment related.
+     */
+    event DisputeOpened(bytes indexed paymentReference);
+
+    /**
+     * @notice Emitted when a dispute has been successfully resolved.
+     * @param paymentReference Reference of the payment related.
+     */
+    event DisputeResolved(bytes indexed paymentReference);
```

Comment:
Same comment as above for the name "Dispute"

Reply:
My thought behind `RequestCompleted(bytes indexed paymentReference);` suggestion is this:

When a `FrozenRequest` is resolved, the function `_withdraw(_paymentRef);` is executed. `

`_withdraw(_paymentRef)` utilizes **ERC20FeeProxy.sol** and its `transferFromWithReferenceAndFee()` function to pay the amount and fee before deleting the `Request`  data from `FrozenRequest` struct and emitting `RequestCompleted (bytes indexed paymentReference);` event at the end. 

Thus, `RequestCompleted(bytes indexed paymentReference)` can be a good name suggestion for this event.

When the blockchain confirms the transaction, this request can then have **`PAID`** status in our frontend.